ext/bcmath/libbcmath/src/div.c

```diff
@@ -31,219 +31,434 @@
+	 * n1_top = 1234, n2_top = 1
+	 * n2_top_tmp = 12345 (+4 digits), n1_top_tmp = 12345678 (+4 digits)
+	 * tmp_quot = n1_top_tmp / n2_top_tmp = 1000, tmp_rem = -9000
+	 * tmp_quot is too large, so tmp_quot--, tmp_rem += n2 (restoring)
+	 * quot = 999, rem = 123447789
+	 *
+	 * Details:
+	 * Suppose that when calculating the temporary quotient, only the high-order elements of the BC_VECTOR array are used.
+	 * At this time, n1 and n2 can be considered to be decomposed as follows. (Here, B = 10^b, any b ∈ ℕ)
+	 * n1 = n1_high * B^k + n1_low
+	 * n2 = n2_high * B^k + n2_low
+	 *
+	 * The error E can be expressed by the following formula.
+	 *
+	 * E = (n1_high * B^k) / (n2_high * B^k) - (n1_high * B^k + n1_low) / (n2_high * B^k + n2_low)
+	 * E = n1_high / n2_high - (n1_high * B^k + n1_low) / (n2_high * B^k + n2_low)
+	 * E = (n1_high * (n2_high * B^k + n2_low) - (n1_high * B^k + n1_low) * n2_high) / (n2_high * (n2_high * B^k + n2_low))
+	 * E = (n1_high * n2_low - n2_high * n1_low) / (n2_high * (n2_high * B^k + n2_low))
+	 *
+	 * Find the error MAX_E when the error E is maximum.
+	 * First, n1_high, which only exists in the numerator, uses its maximum value. Considering carry-back,
+	 * n1_high can be expressed as follows.
+	 * n1_high = n2_high * B
+	 * And n1_low is also present only in the numerator, but since it is a negative number, use the minimum value, i.e. 0.
+	 * n1_low = 0
+	 *
+	 * MAX_E = (n1_high * n2_low - n2_high * n1_low) / (n2_high * (n2_high * B^k + n2_low))
+	 * MAX_E = (n2_high * B * n2_low) / (n2_high * (n2_high * B^k + n2_low))
+	 *
+	 * n2_low uses the maximum value.
+	 * n2_low = B^k - 1
+	 * MAX_E = (n2_high * B * n2_low) / (n2_high * (n2_high * B^k + n2_low))
+	 * MAX_E = (n2_high * B * (B^k - 1)) / (n2_high * (n2_high * B^k + B^k - 1))
+	 *
+	 * n2_high uses the minimum value, but want to see how the number of digits affects the error, so represen
+	 * the minimum value as:
+	 * n2_high = 10^x (any x ∈ ℕ)
+	 * Since B = 10^b, the formula becomes:
+	 * MAX_E = (n2_high * B * (B^k - 1)) / (n2_high * (n2_high * B^k + B^k - 1))
+	 * MAX_E = (10^x * 10^b * ((10^b)^k - 1)) / (10^x * (10^x * (10^b)^k + (10^b)^k - 1))
+	 * MAX_E = (10^(x + b) * (10^(k * b) - 1)) / (10^(2x + k * b) + 10^(x + k * b) + 10^x)
+	 * MAX_E = (10^(x + b) * (10^(k * b) - 1)) / (10^(x + k * b) * (10^x + 1) + 10^x)
+	 *
+	 * Now let's make an approximation. Remove -1 from the numerator and +1 from the denominator. Approximate
+	 * the numerator to be large and the denominator to be small, such that MAX_E is less than this expression.
+	 * MAX_E = (10^(x + b) * (10^(k * b) - 1)) / (10^(x + k * b) * (10^x + 1) + 10^x)
+	 * MAX_E < (10^(x + b) * 10^(k * b)) / (10^(x + k * b) * 10^x + 10^x)
+	 * MAX_E < (10^(x + k + 2b)) / (10^x * (10^(x + k * b) + 1))
+	 *
+	 * Approximate again and remove +1 in the denominator
+	 * MAX_E < 10^(x + k + 2b) / (10^x * (10^(x + k * b) + 1))
+	 * MAX_E < 10^(x + k + 2b) / (10^x * 10^(x + k * b))
```

Comment:
How do you go from this line to line 157?

Reply:
ditto